src/lib.rs

```diff
@@ -447,51 +491,66 @@ impl Traits {
     }
 }
 
-#[proc_macro_error]
-#[proc_macro_attribute]
-pub fn derive_where(
-    attr: proc_macro::TokenStream,
-    item: proc_macro::TokenStream,
-) -> proc_macro::TokenStream {
-    let quote_item: TokenStream = item.clone().into();
-    let mut output = quote! { #quote_item };
+/// Internal derive function for handling errors.
+fn derive_where_internal(attr: TokenStream, item: TokenStream) -> Result<TokenStream> {
+    // The item needs to be added, as it is consumed by the derive.
+    let mut output = quote! { #item };
 
-    let derive_where: DeriveWhere = syn::parse(attr).unwrap();
+    // Parse derive macro.
+    let derive_where: DeriveWhere = syn::parse2(attr)?;
 
+    // Parse item.
     let DeriveInput {
         ident,
         generics,
         data,
         ..
-    } = parse_macro_input!(item);
+    } = syn::parse2(item)?;
 
+    // Every trait needs a separate implementation.
     for trait_ in &derive_where.traits {
-        let body = trait_.generate_body(&ident, &data);
+        let body = trait_.generate_body(&ident, &data)?;
         let trait_ = trait_.type_();
 
-        let bounds = if derive_where.bounds.is_empty() {
-            quote! {}
-        } else {
-            let mut bounds = quote! { where };
+        // Build necessary generics to construct the implementation item.
+        let (impl_generics, type_generics, where_clause) = generics.split_for_impl();
+        // TODO: don't convert to `TokenStream`, but actually properly insert bounds.
+        let mut where_clause = where_clause.map(|where_clause| where_clause.into_token_stream());
+
+        // If there are any bounds, insert them into the `where` clause.
+        if !derive_where.bounds.is_empty() {
+            // If there is no `where` clause, make one.
+            let where_clause = where_clause.get_or_insert(quote! { where });
 
             for bound in &derive_where.bounds {
-                bounds.extend(quote! {
+                where_clause.extend(quote! {
                     #bound: #trait_,
                 })
             }
+        }
```

Comment:
1. Do we need to support empty bounds?
2. I think the DeriveWhere parser fails on empty bounds or traits.

Reply:
1. We do have to support empty bounds:
```rust
struct Test<T>(PhantomData<T>);

impl<T> Clone for Test<T> {
  fn clone(&self) -> Self {
    Self(PhantomData)
  }
}
```
2. Indeed, something we still have to support, should be tracked at #2. We still have to figure out how the syntax will look like then.

Reply:
Well probably just use empty bounds, we could use `look_ahead` and try parse the DeriveWhere and if it fails parse only Trait list.
so `#[derive_where(Clone)]` without `;` and `T`.

Reply:
This is part of #2 and should not block this.